ding/model/template/vae.py

```diff
@@ -0,0 +1,210 @@
+"""Note the following vae model is borrowed from https://github.com/AntixK/PyTorch-VAE"""
+
+import torch
+from torch.nn import functional as F
+from torch import nn
+from abc import abstractmethod
+from typing import List, Callable, Union, Any, TypeVar, Tuple
+
+# from torch import tensor as Tensor
+Tensor = TypeVar('torch.tensor')
+
+
+class BaseVAE(nn.Module):
+
+    def __init__(self) -> None:
+        super(BaseVAE, self).__init__()
+
+    def encode(self, input: Tensor) -> List[Tensor]:
+        raise NotImplementedError
+
+    def decode(self, input: Tensor) -> Any:
+        raise NotImplementedError
+
+    def sample(self, batch_size: int, current_device: int, **kwargs) -> Tensor:
+        raise RuntimeWarning()
+
+    def generate(self, x: Tensor, **kwargs) -> Tensor:
+        raise NotImplementedError
+
+    @abstractmethod
+    def forward(self, *inputs: Tensor) -> Tensor:
+        pass
+
+    @abstractmethod
+    def loss_function(self, *inputs: Any, **kwargs) -> Tensor:
+        pass
+
+
+class VanillaVAE(BaseVAE):
+
+    def __init__(self, action_dim: int, obs_dim: int, latent_dim: int, hidden_dims: List = None, **kwargs) -> None:
+        super(VanillaVAE, self).__init__()
+
+        self.action_dim = action_dim
+        self.obs_dim = obs_dim
+        self.latent_dim = latent_dim
+        self.hidden_dims = hidden_dims
+
+        modules = []
+        if hidden_dims is None:
+            hidden_dims = [256]
+
+        # Build Encoder
+        # action
+        self.encode_action_head = nn.Sequential(nn.Linear(self.action_dim, hidden_dims[0]), nn.ReLU())
```

Comment:
current implementation doesn't take the case `len(hidden_dims) > 1` into consideration, and when it does, it will have no influence for this module.